### ISSUE TITLE
ApplicationSet not generating new Application on webhook with Git file generator

### ISSUE THREAD

**Issue by amitlin** [argoproj/argo-cd#16062]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

https://cloud-native.slack.com/archives/C01U45M2SVB/p1697465637463289

When the applicationset controller receives a webhook, it only refreshes existing applicationsets, it does not rescan the repository for new applications when using the Git file generator.

**To Reproduce**

Add the following applicationset:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: test-service
  namespace: argocd-ops
spec:
  goTemplate: true
  goTemplateOptions: ["missingkey=error"]
  generators:
    - git:
        repoURL: https://github.com/someorg/argocd.git
        revision: HEAD
        files:
          - path: "environments/**/test-service/config.json"
  template:
    metadata:
      name: "{{.namespace}}-{{.serviceName}}"
    spec:
      project: default
      sources:
        - chart: kubernetes-dashboard
          repoURL: https://kubernetes.github.io/dashboard/
          targetRevision: "{{.chartVersion}}"
      destination:
        name: "in-cluster"
        namespace: "{{.namespace}}"
      syncPolicy:
        automated:
          prune: true
          selfHeal: true
        syncOptions:
          - CreateNamespace=true
```

And add the following `config.json` to the source argocd repo described in the applicationset, in the following path - `environments/staging/test-namespace/services/test-service/config.json`:
```json
{
  "serviceName": "test-service",
  "namespace": "test-namespace",
  "env": "staging",
  "chartVersion": "*"
}
```
Push and look at the logs.

**Expected behavior**
The webhook should cause argocd to scan the repo again and to generate missing applications.

**Version**
```shell
argocd: v2.8.1+2bc94af.dirty
  BuildDate: 2023-08-22T23:51:03Z
  GitCommit: 2bc94af7bd081bc4682c2c6dc005f6336c754c2b
  GitTreeState: dirty
  GoVersion: go1.21.0
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.8.4+c279299\
```

**Logs**

```
######## START Webhook logs
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:30:14Z" level=info msg="refresh ApplicationSet argocd-ops/test-service from webhook"
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:30:14Z" level=info msg="generated 0 applications" generator="{nil nil &GitGenerator{RepoURL:https://github.com/someorg/argocd-applicative.git,Directories:[]GitDirectoryGeneratorItem{},Files:[]GitFileGeneratorItem{GitFileGeneratorItem{Path:environments/**/test-service/config.json,},},Revision:HEAD,RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},PathParamPrefix:,Values:map[string]string{},} nil nil nil nil nil nil nil}"
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:30:14Z" level=info msg="end reconcile" applicationset=argocd-ops/test-service requeueAfter=3m0s
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:30:14Z" level=info msg="generated 0 applications" generator="{nil nil &GitGenerator{RepoURL:https://github.com/someorg/argocd-applicative.git,Directories:[]GitDirectoryGeneratorItem{},Files:[]GitFileGeneratorItem{GitFileGeneratorItem{Path:environments/**/test-service/config.json,},},Revision:HEAD,RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},PathParamPrefix:,Values:map[string]string{},} nil nil nil nil nil nil nil}"
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:30:14Z" level=info msg="end reconcile" applicationset=argocd-ops/test-service requeueAfter=3m0s
######## END Webhook logs

######## START Interval scan logs
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="generated 1 applications" generator="{nil nil &GitGenerator{RepoURL:https://github.com/someorg/argocd-applicative.git,Directories:[]GitDirectoryGeneratorItem{},Files:[]GitFileGeneratorItem{GitFileGeneratorItem{Path:environments/**/test-service/config.json,},},Revision:HEAD,RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},PathParamPrefix:,Values:map[string]string{},} nil nil nil nil nil nil nil}"
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="created Application" app=test-namespace-test-service appSet=test-service
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="end reconcile" applicationset=argocd-ops/test-service requeueAfter=3m0s
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="generated 1 applications" generator="{nil nil &GitGenerator{RepoURL:https://github.com/someorg/argocd-applicative.git,Directories:[]GitDirectoryGeneratorItem{},Files:[]GitFileGeneratorItem{GitFileGeneratorItem{Path:environments/**/test-service/config.json,},},Revision:HEAD,RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},PathParamPrefix:,Values:map[string]string{},} nil nil nil nil nil nil nil}"
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="unchanged Application" app=test-namespace-test-service appSet=test-service
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="end reconcile" applicationset=argocd-ops/test-service requeueAfter=3m0s
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="generated 1 applications" generator="{nil nil &GitGenerator{RepoURL:https://github.com/someorg/argocd-applicative.git,Directories:[]GitDirectoryGeneratorItem{},Files:[]GitFileGeneratorItem{GitFileGeneratorItem{Path:environments/**/test-service/config.json,},},Revision:HEAD,RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},PathParamPrefix:,Values:map[string]string{},} nil nil nil nil nil nil nil}"
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="unchanged Application" app=test-namespace-test-service appSet=test-service
argocd-applicationset-controller-56664994d5-bszbt applicationset-controller time="2023-10-22T07:31:30Z" level=info msg="end reconcile" applicationset=argocd-ops/test-service requeueAfter=3m0s
######## END Interval scan logs
```

We can see that the webhook itself did not cause the application to be generated, and only the interval scan about a minute or so later caused it to be generated.

**Comment by Laakso** [comment]
We are seeing exactly the same behaviour.

**Comment by dhruvang1** [comment]
I am working on the fix in #16241
